### ISSUE TITLE
Potential infinite node visit loop

### ISSUE THREAD

**Issue by longlho** [babel/babel#12976]
## Bug Report

**Current behavior**

Right now a plugin processing a node type can get into an infinite loop if it tries to insert another node of the same type.

If I want to make a node as visited, I can either use a `WeakSet` or to set a private attribute on that node. However, other plugins can copy the node I visited and move it out of scope and do w/e that would potentially trigger my plugin again, resulting in a potential infinite loop.

My immediate problem is that babel-plugin-formatjs cannot process a `CallExpression` it has already modified and right now I can't think of a way to safely mark it as processed such that downstream plugin cannot mess with it in a way that triggers my plugin again.

**Input Code**

https://astexplorer.net/#/gist/13be93f1a4d27755a59e6cea302787b0/681287d42012702a8a1afcda569556a263f9310c

**Expected behavior**

Once a node has been visited by a plugin it shouldn't be visited by the same plugin

**Possible Solution**

**Additional context**

Upstream issue: https://github.com/formatjs/formatjs/issues/2663

**Comment by conartist6** [comment]
Your example doesn't make a lot of sense to me. It seems to me you have a few options.

One, you could separate your plugin out into its own pass. This obviously has a major perf cost -- it can double the time required for transpilation -- which may or may not be significant for you. You can lower the cost by running babel programmatically and using the [transformFromAst](https://babeljs.io/docs/en/babel-core#transformfromast) method so that parsing does not happen twice.

Alternatively you could embrace babel the way it is meant to work. This means altering your code so that it is able to recognize a spot where no further transformation is needed, thus averting an infinite loop. In the case of your example this would be as simple as checking whether the expression is already preceded by the code your plugin inserts.

I would recommend against adding properties to the AST nodes themselves, though it certainly could be used to do what you describe.

**Comment by longlho** [comment]
Thanks for the response. Can you clarify the statement below?

> In the case of your example this would be as simple as checking whether the expression is already preceded by the code your plugin inserts.

If a plugin makes a copy of my already processed node, how can I detect that? Or rather, how do I safely mark a node as processed? I've tried setting an attribute on the Node itself and that doesn't work, the only alternative is to add an argument, or a key in the argument node (if it's a `CallExpression`) but that seems suboptimal.

**Comment by conartist6** [comment]
Could you clarify what your example code is meant to do? It would help me explain what I think you should do. I broke the infinite recursion manually (not something that will help you), and the output did not make much sense to me:

https://astexplorer.net/#/gist/13be93f1a4d27755a59e6cea302787b0/910b45cbb634a2d13f5b8bf6a620d9b88a4b337e

**Comment by longlho** [comment]
Sorry the example code is super watered down. The plugin I wrote is rather large but at a high level it does something like this: https://astexplorer.net/#/gist/13be93f1a4d27755a59e6cea302787b0/7fb4f0846bba64d6f121f3f69244c3b017f1472c

So basically if a processed node is processed twice it'll bomb out. The specific issue I'm seeing is that transformation for `async` function seems to copy the processed node, thus throwing an exception in my plugin.

**Comment by conartist6** [comment]
Ah, you're extracting translations or errors codes or something I assume. I believe what you really want to do is to ensure that this processing happens before any of the language transpilation does (so that its input is the code as it was written). Babel usually does a single traversal mixing the node handlers defined by different plugins. The mixing is causing your problem. To avoid it all you need to do is trigger your own traversal:

```js
import traverse from "@babel/traverse";

export default function (babel) {
  const { types: t } = babel;

  return {
    visitor: {
      Program(path) {
        traverse(path.node, {
          CallExpression(path) {
            // ...
          }
        })
      }
    }
  };
}
```

**Comment by conartist6** [comment]
`babel-plugin-macros` does the same thing to avoid that problem.

**Comment by longlho** [comment]
gotcha thanks a lot for the pointer I'll give that a try!

**Comment by conartist6** [comment]
I believe it's not the way they hope to handle this issue in the long run. For you it's fine, and will continue to be, since your work really should always happen first. But some transpilation plugins have similar issues but can't afford to make such assumptions. So there is [an alternate solution under development](https://github.com/babel/babel/issues/5854) which I presume will be promoted and documented fully once it is complete. I notice though that that issue links an article which has a lot of detail on this issue: https://jamie.build/babel-plugin-ordering.html

**Comment by longlho** [comment]
Closing since there's a workaround. Thanks for the help!
